File: src/text/text-layout.ts
import type { Font } from '../font/metrics.js';
import type { TextStyle, Alignment, RichTextRun, RichTextOptions } from './text-style.js';

export interface LayoutWord {
  text: string;
  width: number;
}

export interface LayoutLine {
  text: string;
  width: number;
  x: number;
  y: number;
  words: LayoutWord[];
}

export interface LayoutResult {
  lines: LayoutLine[];
  totalHeight: number;
  totalWidth: number;
}

export interface RichLayoutRun {
  text: string;
  width: number;
  font: Font;
  fontSize: number;
  color?: import('../color/color.js').Color;
  underline?: boolean;
  strikethrough?: boolean;
  link?: string;
  x: number; // x offset within the line
}

export interface RichLayoutLine {
  runs: RichLayoutRun[];
  width: number;
  x: number;
  y: number;
  height: number; // max line height across runs
}

export interface RichLayoutResult {
  lines: RichLayoutLine[];
  totalHeight: number;
  totalWidth: number;
}

/**
 * Split text into words, preserving explicit newlines as separate tokens.
 * A newline token is represented as '\n'.
 */
function splitIntoWords(text: string): string[] {
  const result: string[] = [];
  const segments = text.split('\n');
  for (let i = 0; i < segments.length; i++) {
    if (i > 0) {
      result.push('\n');
    }
    const words = segments[i].split(/\s+/).filter(w => w.length > 0);
    for (const word of words) {
      result.push(word);
    }
  }
  return result;
}

/**
 * Measure a single word width including optional letter spacing.
 */
function measureWord(font: Font, fontSize: number, word: string, letterSpacing: number): number {
  const baseWidth = font.measureWidth(word, fontSize);
  // Letter spacing applies between characters, so (chars - 1) extra spaces
  if (letterSpacing !== 0 && word.length > 1) {
    return baseWidth + letterSpacing * (word.length - 1);
  }
  return baseWidth;
}

/**
 * Measure a space character width.
 */
function measureSpace(font: Font, fontSize: number, wordSpacing: number): number {
  return font.measureWidth(' ', fontSize) + wordSpacing;
}

/**
 * Layout text into lines respecting maxWidth and maxLines constraints.
 */
export function layoutText(
  text: string,
  style: TextStyle,
  maxWidth?: number,
  maxLines?: number,
): LayoutResult {
  const { font, fontSize, alignment = 'left', lineHeight: lineHeightMultiplier = 1.2 } = style;
  const letterSpacing = style.letterSpacing ?? 0;
  const wordSpacing = style.wordSpacing ?? 0;

  const lineHeight = fontSize * lineHeightMultiplier;
  const words = splitIntoWords(text);
  const spaceWidth = measureSpace(font, fontSize, wordSpacing);

  const lines: LayoutLine[] = [];
  let currentWords: LayoutWord[] = [];
  let currentWidth = 0;

  function finishLine(): void {
    if (maxLines !== undefined && lines.length >= maxLines) {
      return;
    }

    const lineText = currentWords.map(w => w.text).join(' ');
    const lineWidth = currentWidth;
    currentWords = [];
    currentWidth = 0;

    lines.push({
      text: lineText,
      width: lineWidth,
      x: 0, // will be adjusted for alignment
      y: 0, // will be set below
      words: [],
    });
  }

  for (const word of words) {
    if (word === '\n') {
      // Explicit newline: finish current line
      finishLine();
      if (maxLines !== undefined && lines.length >= maxLines) break;
      continue;
    }

    const wordWidth = measureWord(font, fontSize, word, letterSpacing);
    const widthWithWord = currentWords.length === 0
      ? wordWidth
      : currentWidth + spaceWidth + wordWidth;

    if (maxWidth !== undefined && currentWords.length > 0 && widthWithWord > maxWidth) {
      // Word exceeds max width, wrap to next line
      finishLine();
      if (maxLines !== undefined && lines.length >= maxLines) break;
    }

    if (currentWords.length === 0) {
      currentWidth = wordWidth;
    } else {
      currentWidth += spaceWidth + wordWidth;
    }
    currentWords.push({ text: word, width: wordWidth });
  }

  // Finish last line if there are remaining words
  if (currentWords.length > 0) {
    finishLine();
  }

  // If there were no words but text was empty or only whitespace, produce one empty line
  if (lines.length === 0 && text.length > 0) {
    lines.push({
      text: '',
      width: 0,
      x: 0,
      y: 0,
      words: [],
    });
  }

  // Assign y positions (PDF y goes up, so first line is at top, subsequent lines go down)
  // y=0 is the baseline of the first line; subsequent lines are offset negatively
  let totalWidth = 0;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    line.y = i === 0 ? 0 : -(i * lineHeight);

    // Recalculate words for the line
    const lineWords = line.text.split(' ').filter(w => w.length > 0);
    line.words = lineWords.map(w => ({
      text: w,
      width: measureWord(font, fontSize, w, letterSpacing),
    }));

    // Alignment offset
    if (maxWidth !== undefined) {
      switch (alignment) {
        case 'center':
          line.x = (maxWidth - line.width) / 2;
          break;
        case 'right':
          line.x = maxWidth - line.width;
          break;
        case 'justify':
          line.x = 0; // justify handled via word spacing in renderer
          break;
        case 'left':
        default:
          line.x = 0;
          break;
      }
    }

    if (line.width > totalWidth) {
      totalWidth = line.width;
    }
  }

  const totalHeight = lines.length > 0 ? (lines.length - 1) * lineHeight + fontSize : 0;

  return { lines, totalHeight, totalWidth };
}

/**
 * Layout rich text (mixed fonts/sizes) into lines.
 */
export function layoutRichText(
  runs: RichTextRun[],
  options: RichTextOptions,
  defaultFont: Font,
  defaultFontSize: number,
): RichLayoutResult {
  const { width, alignment = 'left', lineHeight: lineHeightMultiplier = 1.2, maxLines } = options;

  // First, flatten all runs into word-level fragments with their style info
  interface StyledWord {
    text: string;
    font: Font;
    fontSize: number;
    color?: import('../color/color.js').Color;
    underline?: boolean;
    strikethrough?: boolean;
    link?: string;
    width: number;
    spaceWidth: number;
    lineHeight: number;
    isNewline: boolean;
  }

  const styledWords: StyledWord[] = [];

  for (const run of runs) {
    const font = run.font ?? defaultFont;
    const fontSize = run.fontSize ?? defaultFontSize;
    const lh = fontSize * lineHeightMultiplier;
    const sw = font.measureWidth(' ', fontSize);

    const segments = run.text.split('\n');
    for (let si = 0; si < segments.length; si++) {
      if (si > 0) {
        styledWords.push({
          text: '\n',
          font,
          fontSize,
          color: run.color,
          underline: run.underline,
          strikethrough: run.strikethrough,
          link: run.link,
          width: 0,
          spaceWidth: sw,
          lineHeight: lh,
          isNewline: true,
        });
      }
      const words = segments[si].split(/\s+/).filter(w => w.length > 0);
      for (const word of words) {
        styledWords.push({
          text: word,
          font,
          fontSize,
          color: run.color,
          underline: run.underline,
          strikethrough: run.strikethrough,
          link: run.link,
          width: font.measureWidth(word, fontSize),
          spaceWidth: sw,
          lineHeight: lh,
          isNewline: false,
        });
      }
    }
  }

  // Break into lines
  const lines: RichLayoutLine[] = [];
  let currentLineWords: StyledWord[] = [];
  let currentLineWidth = 0;

  function finishRichLine(): void {
    if (maxLines !== undefined && lines.length >= maxLines) return;
    if (currentLineWords.length === 0) {
      // Empty line (from explicit newline)
      const defaultLH = defaultFontSize * lineHeightMultiplier;
      lines.push({
        runs: [],
        width: 0,
        x: 0,
        y: 0,
        height: defaultLH,
      });
      currentLineWords = [];
      currentLineWidth = 0;
      return;
    }

    // Build runs for this line, merging consecutive words with same style
    const lineRuns: RichLayoutRun[] = [];
    let xOffset = 0;
    let maxHeight = 0;

    for (let i = 0; i < currentLineWords.length; i++) {
      const w = currentLineWords[i];
      if (w.lineHeight > maxHeight) maxHeight = w.lineHeight;

      // Add space before word if not the first
      if (i > 0) {
        xOffset += currentLineWords[i - 1].spaceWidth;
      }

      lineRuns.push({
        text: w.text,
        width: w.width,
        font: w.font,
        fontSize: w.fontSize,
        color: w.color,
        underline: w.underline,
        strikethrough: w.strikethrough,
        link: w.link,
        x: xOffset,
      });

      xOffset += w.width;
    }

    lines.push({
      runs: lineRuns,
      width: currentLineWidth,
      x: 0,
      y: 0,
      height: maxHeight,
    });

    currentLineWords = [];
    currentLineWidth = 0;
  }

  for (const sw of styledWords) {
    if (sw.isNewline) {
      finishRichLine();
      if (maxLines !== undefined && lines.length >= maxLines) break;
      continue;
    }

    const widthWithWord = currentLineWords.length === 0
      ? sw.width
      : currentLineWidth + currentLineWords[currentLineWords.length - 1].spaceWidth + sw.width;

    if (currentLineWords.length > 0 && widthWithWord > width) {
      finishRichLine();
      if (maxLines !== undefined && lines.length >= maxLines) break;
    }

    if (currentLineWords.length === 0) {
      currentLineWidth = sw.width;
    } else {
      currentLineWidth += currentLineWords[currentLineWords.length - 1].spaceWidth + sw.width;
    }
    currentLineWords.push(sw);
  }

  if (currentLineWords.length > 0) {
    finishRichLine();
  }

  // Set y positions and alignment
  let yPos = 0;
  let totalWidth = 0;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    line.y = -yPos;
    yPos += line.height;

    // Alignment
    switch (alignment) {
      case 'center':
        line.x = (width - line.width) / 2;
        break;
      case 'right':
        line.x = width - line.width;
        break;
      case 'justify':
      case 'left':
      default:
        line.x = 0;
        break;
    }

    if (line.width > totalWidth) totalWidth = line.width;
  }

  const totalHeight = yPos;

  return { lines, totalHeight, totalWidth };
}

File: src/text/text_layout.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const StandardFont = @import("../font/standard_fonts.zig").StandardFont;

/// A single laid-out line of text with its measured width.
pub const TextLine = struct {
    /// The text content of this line (a slice into the original string).
    text: []const u8,
    /// The measured width of this line in PDF points.
    width: f32,
    /// Byte offset of the start of this line in the original text.
    start: usize,
    /// Byte offset of the end of this line in the original text (exclusive).
    end: usize,
};

/// Lays out text into lines, splitting on whitespace and wrapping at max_width.
/// Handles embedded newlines as forced line breaks.
/// Caller owns the returned slice and must free it with the same allocator.
pub fn layoutText(
    allocator: Allocator,
    text: []const u8,
    font: StandardFont,
    font_size: f32,
    max_width: ?f32,
) ![]TextLine {
    var lines: ArrayList(TextLine) = .{};
    errdefer lines.deinit(allocator);

    if (text.len == 0) {
        return lines.toOwnedSlice(allocator);
    }

    // Split on newlines first, then wrap each paragraph.
    var para_start: usize = 0;
    while (para_start <= text.len) {
        // Find the end of this paragraph (next newline or end of text).
        var para_end = para_start;
        while (para_end < text.len and text[para_end] != '\n') {
            para_end += 1;
        }

        const paragraph = text[para_start..para_end];

        if (paragraph.len == 0) {
            // Empty paragraph (consecutive newline or trailing newline).
            try lines.append(allocator, .{
                .text = paragraph,
                .width = 0,
                .start = para_start,
                .end = para_end,
            });
        } else {
            // Wrap this paragraph.
            try wrapParagraph(allocator, &lines, paragraph, para_start, font, font_size, max_width);
        }

        para_start = para_end + 1;
    }

    return lines.toOwnedSlice(allocator);
}

/// Wraps a single paragraph (no embedded newlines) into lines.
fn wrapParagraph(
    allocator: Allocator,
    lines: *ArrayList(TextLine),
    paragraph: []const u8,
    base_offset: usize,
    font: StandardFont,
    font_size: f32,
    max_width: ?f32,
) !void {
    const space_width = font.textWidth(" ", font_size);
    const limit = max_width orelse std.math.floatMax(f32);

    var line_start: usize = 0;
    var line_width: f32 = 0;
    var word_start: usize = 0;
    var i: usize = 0;

    // Skip leading spaces.
    while (i < paragraph.len and paragraph[i] == ' ') {
        i += 1;
    }
    line_start = i;
    word_start = i;

    while (i <= paragraph.len) {
        // Find word boundaries.
        if (i == paragraph.len or paragraph[i] == ' ') {
            if (word_start < i) {
                const word = paragraph[word_start..i];
                const word_width = font.textWidth(word, font_size);

                if (line_width > 0 and line_width + space_width + word_width > limit) {
                    // Wrap: emit current line, start new one with this word.
                    var trimmed_end = word_start;
                    while (trimmed_end > line_start and paragraph[trimmed_end - 1] == ' ') {
                        trimmed_end -= 1;
                    }
                    const line_text = paragraph[line_start..trimmed_end];
                    const measured = font.textWidth(line_text, font_size);
                    try lines.append(allocator, .{
                        .text = line_text,
                        .width = measured,
                        .start = base_offset + line_start,
                        .end = base_offset + trimmed_end,
                    });
                    line_start = word_start;
                    line_width = word_width;
                } else {
                    if (line_width > 0) {
                        line_width += space_width;
                    }
                    line_width += word_width;
                }
            }
            // At end of paragraph, break out of the loop.
            if (i == paragraph.len) break;
            // Skip spaces.
            while (i < paragraph.len and paragraph[i] == ' ') {
                i += 1;
            }
            word_start = i;
        } else {
            i += 1;
        }
    }

    // Emit the last line.
    var trimmed_end = paragraph.len;
    while (trimmed_end > line_start and paragraph[trimmed_end - 1] == ' ') {
        trimmed_end -= 1;
    }
    if (trimmed_end > line_start or lines.items.len == 0) {
        const line_text = paragraph[line_start..trimmed_end];
        const measured = font.textWidth(line_text, font_size);
        try lines.append(allocator, .{
            .text = line_text,
            .width = measured,
            .start = base_offset + line_start,
            .end = base_offset + trimmed_end,
        });
    }
}

/// Calculates the total height needed to render the given lines.
pub fn measureTextHeight(lines: []const TextLine, font_size: f32, line_height: f32) f32 {
    if (lines.len == 0) return 0;
    _ = font_size;
    return @as(f32, @floatFromInt(lines.len)) * line_height;
}

// -- Tests --

test "layout single line no wrap" {
    const lines = try layoutText(std.testing.allocator, "Hello World", .helvetica, 12.0, null);
    defer std.testing.allocator.free(lines);

    try std.testing.expectEqual(@as(usize, 1), lines.len);
    try std.testing.expectEqualStrings("Hello World", lines[0].text);
}

test "layout empty text" {
    const lines = try layoutText(std.testing.allocator, "", .helvetica, 12.0, null);
    defer std.testing.allocator.free(lines);
    try std.testing.expectEqual(@as(usize, 0), lines.len);
}

test "layout with newlines" {
    const lines = try layoutText(std.testing.allocator, "Line 1\nLine 2\nLine 3", .helvetica, 12.0, null);
    defer std.testing.allocator.free(lines);

    try std.testing.expectEqual(@as(usize, 3), lines.len);
    try std.testing.expectEqualStrings("Line 1", lines[0].text);
    try std.testing.expectEqualStrings("Line 2", lines[1].text);
    try std.testing.expectEqualStrings("Line 3", lines[2].text);
}

test "layout with wrapping" {
    // "Hello" in Helvetica at 12pt = (722+556+222+222+556)/1000*12 = ~27.3pt
    // "World" = (944+556+333+222+500)/1000*12 = ~30.7pt
    // space = 278/1000*12 = ~3.3pt
    // Total ~61.3pt. With max_width=50, should wrap.
    const lines = try layoutText(std.testing.allocator, "Hello World", .helvetica, 12.0, 50.0);
    defer std.testing.allocator.free(lines);

    try std.testing.expectEqual(@as(usize, 2), lines.len);
    try std.testing.expectEqualStrings("Hello", lines[0].text);
    try std.testing.expectEqualStrings("World", lines[1].text);
}

test "measure text height" {
    const dummy = [_]TextLine{
        .{ .text = "a", .width = 10, .start = 0, .end = 1 },
        .{ .text = "b", .width = 10, .start = 2, .end = 3 },
        .{ .text = "c", .width = 10, .start = 4, .end = 5 },
    };
    const height = measureTextHeight(&dummy, 12.0, 14.4);
    try std.testing.expectApproxEqAbs(@as(f32, 43.2), height, 0.1);
}
